src/main/java/fi/aalto/cs/apluscourses/utils/FactorySelector.java

```diff
@@ -0,0 +1,36 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class FactorySelector<A, R> {
+  private final ConcurrentHashMap<Class<?>, Factory<?, A, R>> factories = new ConcurrentHashMap<>();
+
+  public <T> void register(@NotNull Class<T> klass, @NotNull Factory<? super T, A, R> factory) {
+    factories.put(klass, factory);
+  }
+
+  @NotNull
+  public <T> R create(T obj, A arg) {
+    return Optional.ofNullable(createInternal(obj.getClass(), obj, arg))
+        .orElseThrow(() -> new IllegalArgumentException("Cannot find suitable factory"));
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> R createInternal(@Nullable Class<? extends T> klass, T obj, A arg) {
+    if (klass == null) {
+      return null;
+    }
+    Factory<T, A, R> factory = (Factory<T, A, R>) factories.get(klass);
+    return factory == null
+        ? createInternal(klass.getSuperclass(), obj, arg)
+        : factory.create(obj, arg);
+  }
+
+  public interface Factory<T, A, R> {
+    @NotNull
+    R create(@NotNull T model, @Nullable A arg);
+  }
+}
```

Comment:
This class adds a lot of complexity compared to the benefits it brings (as far as I can see). Personally, I would really prefer a simpler approach even if it means some code duplication etc. :smile:

Reply:
Hmm... I was just thinking how easily it would solve an issue I was facing (well, not really facing but more like thinking of) with `Authentication`.

So, for integration tests (and in the future in production, too), we need to use different authentication method. The most conventional way to implement that is having two subclasses of `Authentication`: `TokenAuthentication` and `PasswordAuthentication`. When we'll support password authentication also in the UI, we'll have `PasswordAuthenticationViewModel` (`TokenAuthenticationViewModel` we already have, though it's differently named). We would also have separate views (dialogs) for these.

In this case, we could just set up
```java
Dialogs.register(TokenAuthenticationViewModel.class, TokenAuthenticationDialog::new);
Dialogs.register(PasswordAuthenticationViewModel.class, PasswordAuthenticationDialog::new);
```
and having an `AuthenticationViewModel` (not knowing is it token or password), we could call
```java
Dialogs.create(authenticationViewModel).showAndGet()
```
to show the correct kind of dialog to the user and consider authentication configured.

So I like this class but I think we can drop it if you want to.

Reply:
That makes sense, but when we are showing the authentication dialog, we don't (I think) really care what the current authentication method is, instead we care how the use wants to authenticate _now_. Maybe the different authentication methods have different menu items or there is some other way for the user to choose, but when the authentication dialog is shown, we should already know which one we are showing.

Then again showing the same type of dialog as the existing one could be useful if the authentication is incorrect and we want to prompt the user to re-enter it so it could be useful now that I think about it

Reply:
Although even if we prompt the user to re-authenticate, it may make sense to let them select between token/user+pw

Reply:
You're correct. Well, the authentication was more like an example for a case where this class could be useful. Selector mechanisms that choose a view based on view model type are, however, typical for UI frameworks following the presentation model (MVVM), so this is somewhat of a common pattern.

Reply:
Hmm, perhaps this fits into the "I don't know how it's implemented but I can still use it" -category that you mentioned on Friday, but for that to actually be true, it could at least use some javadocs.

As far as common MVVM patterns goes, I have no idea and defer to your knowledge there :smile:

Reply:
and here are my two cents...

> Hmm... I was just thinking how easily it would solve an issue I was facing (well, not really facing but more like thinking of) with `Authentication`.

I strongly suggest we solve issues **a)** that exist, and **b)** when they appear - the whole point of agile is to deliver "business value" quickly. as an approach, it is totally allowing duplications/repetitions etc. as long as there is a benefit for the "value". it does not really cover the architectural side of things, but from the business POV, there is no benefit of a good architecture if you've lost the competition to the crappy-but-quickly-developed-service.



> This class adds a lot of complexity compared to the benefits it brings (as far as I can see). Personally, I would really prefer a simpler approach even if it means some code duplication etc.

here and in some other places, there are quite complex to perceive generic structures already introduced. They are nice and elegant, but again, **untested**. And eventually, there is no resource for testing them atm (and probably never in the future), not to mention future maintenance burden and being locked mostly to one team member availability. Repeating myself, I'd rather **a)** go with a stupid `for-loop` as a trivial (but no tests needed solution) or some kind of a **b)** library/third-party solution, that has been used 1kk times elsewhere, stable and "not-my-problem".

another point is, that things, for them to be used, should be simple and understandable. in other words, there's little benefit if 2/3 of the code uses the "good-old" `for-loop` approach and one does it with a fancy one...

> So, for integration tests (and in the future in production, too), we need to use different authentication method.

as far as I remember the **RestAssured** might have also some way to auth with token too, but here the "purpose" of testing is a bit exaggerated I think. as long as we can make sure, that for a certain "user" we can get the data we need from A+ -> we're good. the _raison d'etre_ for the integration tests is a bit unconventional...

Reply:
Pourquoi utilisez-vous le français?

Reply:
Excellent points. A few notes:
> I'd rather a) go with a stupid for-loop as a trivial (but no tests needed solution)

I don't see how inlining `for` loops would decrease the need for testing, quite the contrary. (I assume we are talking about loops like `for (int i = 0; i < length; i++) { ... }`. ) I actually think that even small amount of additional logic included in the same method with a `for` loop increases its complexity quickly to a level where the method is no longer (thoroughly) testable. Knowing how `for` loops are often sources of bugs (off-by-one errors, unintentional closures, illegal modifications during iteration, ...) I'd avoid them, if possible, in business logic code.

>b) library/third-party solution, that has been used 1kk times elsewhere, stable and "not-my-problem"

Each library dependency is a real maintenance burden, IMO. This of course doesn't mean libraries shouldn't be used, but the design decision on which libraries are used, by which parts of the code and how, must be made with a careful consideration. Using libraries incautiously makes them easily spread from the parts of the system where they're actually needed to all around the codebase. For better maintainability, I'd like to keep library-reliant code behind [bridges](https://en.wikipedia.org/wiki/Bridge_pattern) or [facades](https://en.wikipedia.org/wiki/Facade_pattern) or such whenever possible.